### PR TITLE
TST: use fixtures for some sql database connections

### PR DESCRIPTION
**Pull request by fangchenli**
xref #40686

This PR:

1. Create fixtures for all supported SQL connections, including sqlalchemy engine/connection and sqlite3 connection. Setup and cleanup are handled within fixtures.
2. Migrate several common tests to use connection fixtures. They serve as examples for future migrations.

We probably should put all the fixtures in a separate conftest.py file. But since there are unmerged PRs related to this test file, I kept everything in one place for now to avoid conflict.

### CODE REVIEW

**Comment by mroeschke** [comment]
Thanks for the PR but appears that it has gone stale. Feel free to reopen if you have time to continue working on it!

File: pandas/tests/io/test_sql.py [hunk 1]
```diff
@@ -367,6 +375,270 @@ def test_frame3():
     return DataFrame(data, columns=columns)
 
 
+@pytest.fixture
```
Comment (mroeschke) [inline pandas/tests/io/test_sql.py:378]: Does it makes sense for the the engine and connection fixtures to be `scope="module"` so they are not repeatedly created and destroyed?

File: pandas/tests/io/test_sql.py [hunk 2]
```diff
@@ -367,6 +375,270 @@ def test_frame3():
...
+
+@pytest.fixture
+def mysql_pymysql_conn(mysql_pymysql_engine):
+    yield mysql_pymysql_engine.connect()
+
+
+@pytest.fixture
+def postgresql_psycopg2_engine(iris_path, types_data):
+    sqlalchemy = pytest.importorskip("sqlalchemy")
+    _ = pytest.importorskip("psycopg2")
```
Comment (mroeschke) [inline pandas/tests/io/test_sql.py:409]: Nit: I think the `_ =` can be removed if you essentially want to check for the import

**Comment by jreback** [comment]
@fangchenli happyt o take this if you can fixup

File: pandas/tests/io/test_sql.py (hunk 1 above)
Comment (fangchenli) [reply pandas/tests/io/test_sql.py:378]: The fixtures also do cleanups after each test. So it's better to have a new engine/conn for each test. Also, they don't seem to add much overhead.

**Review by mroeschke — APPROVED** [review]

**Comment by jreback** [comment]
thanks @fangchenli

**Pull request merged**